### PR TITLE
PAPP-29296: Archer - Multi-session issue and feature enhancements

### PR DESCRIPTION
**Pull request by pdesai-crest**
Please ensure your pull request (PR) adheres to the following guidelines:

- Please refer to our contributing documentation for any questions on submitting a pull request, link: [Contribution Guide](https://github.com/splunk-soar-connectors/.github/blob/main/.github/CONTRIBUTING.md)

#### Please check if your PR fulfills the following requirements:
- [ ] Testing of all the changes has been performed (for bug fixes / features)
- [ ] The manual_readme_content.md has been reviewed and added / updated if needed (for bug fixes / features)
- [ ] Use the following format for the PR description: `<App Name>: <PR Type> - <PR Description>`
- [ ] Provide release notes as part of the PR submission which describe high level points about the changes for the upcoming GA release.
- [ ] Verify all checks are passing.
- [ ] Do NOT use the `next` branch of the forked repo. Create separate feature branch for raising the PR.
- [ ] Do NOT submit updates to dependencies unless it fixes an issue.
#### Please check the type of change your PR introduces:
- [ ] New App
- [ ] Bugfix
- [ ] Feature
- [ ] Code style update (formatting, renaming)
- [ ] Refactoring (no functional changes, no api changes)
- [ ] Documentation
- [ ] Other (please describe):

## Security Considerations (REQUIRED)
- [ ] If you are exposing any endpoints using a [REST handler](https://docs.splunk.com/Documentation/SOAR/current/DevelopApps/RESTHandlers),
  please document them in the `manual_readme_content.md`.
- [ ] If this is a new connector or you are adding new actions
    - [ ] Please document in the `manual_readme_content.md` all methods (eg, OAuth) used to authenticate
      with the service that the connector is integrating with.
    - [ ] If any actions are unable to run on SOAR Cloud, please document this in the `manual_readme_content.md`.
- [ ] Are you introducing any new cryptography modules? If yes, please elaborate their purpose:
- [ ] Are you are accessing the file system? If yes, please verify that you are only accessing paths returned through
the [Vault](https://docs.splunk.com/Documentation/SOARonprem/5.2.1/DevelopApps/AppDevAPIRef#Vault) API.
- [ ] Are you are marking code to be ignored by Semgrep with [`nosemgrep`](https://semgrep.dev/docs/ignoring-files-folders-code/#ignoring-code-through-nosemgrep)?
If yes, please provide justification in an additional comment next to the ignored code.

## Release Notes (REQUIRED)
- Provide release notes as part of the PR submission which describe high level points about the changes for the upcoming GA release.

## What is the current behavior? (OPTIONAL)
- Describe the current behavior that you are modifying.

## What is the new behavior? (OPTIONAL)
- Describe the behavior or changes that are being added by this PR.

## Other information (OPTIONAL)
- Any other information that is important to this PR such as screenshots of how the component looks before and after the change.

## Pay close attention to (OPTIONAL)
- Any specific code change or test case points which must be addressed/reviewed at the time of GA release.

## Screenshots (if relevant)

---
Thanks for contributing!

### CODE REVIEW

File: archer_utils.py [hunk 1]
```diff
@@ -683,32 +701,255 @@ def update_record(self, app, contentId, fieldId, value, doit=True):
...
+            W('Updating to type: {}'.format(fieldType))
+            fields.append(field)
+        data = self.asoap.update_record(contentId, moduleId, fields)
+        return bool(data)
+
+    def get_report_by_id(self, guid, max_count, max_pages):
+        """Returns the report with the given guid."""
+
+        # Initialize result dictionary
+        result_dict = {}
```
Comment (bb-splunk) [inline archer_utils.py:760]: Consider:
`result_dict = {'status': 'failed', 'message': 'Failed - default message', 'page_count': 0, 'records': []}`

File: archer_views.py [hunk 2]
```diff
@@ -74,3 +74,44 @@ def list_tickets(provides, all_results, context):
...
+
+    headers = ['content id']
+    context['results'] = results = []
+
+    headers_set = set()
+    for summary, action_results in all_results:
+        for result in action_results:
+            for record in result.get_data():
+                headers_set.update([f.get('@name', '').strip()
+                                    for f in record.get('Field', [])])
```
Comment (bb-splunk) [inline archer_views.py:89]:
```suggestion
                headers_set.update(f.get('@name', '').strip()
                                    for f in record.get('Field', []))
```

**Review by bb-splunk — CHANGES_REQUESTED** [review]

File: archer_utils.py [hunk 3]
```diff
@@ -683,32 +701,255 @@ def update_record(self, app, contentId, fieldId, value, doit=True):
...
+                # Try to get current report page
+                try:
+                    data_dict = self.asoap.get_report(guid, page_number)
+                    if data_dict['status'] != 'success':
+                        result_dict['message'] = data_dict['result']
+                        return result_dict
+                    data = data_dict['result']
+
+                except Exception as e:
+                    result_dict['message'] = 'Failed to get page {} of report. Check input parameters are valid. e = {}'.format(page_number, e)
```
Comment (alexa-phantom) [inline archer_utils.py:784]: Better to use the word error instead of just e here since it is the message to the user

Comment (alexa-phantom) [reply archer_utils.py:784]: Same with other occurrences

**Review by alexa-phantom — CHANGES_REQUESTED** [review]
Would be nice to have consistency with string creation. ie `f"` vs `.format()` - just use one throughout code

File: archer_connector.py [hunk 4]
```diff
@@ -318,144 +335,133 @@ def _handle_create_ticket(self, action_result, param):
...
+            msg = consts.ARCHER_ERR_VALID_JSON
             self.debug_print(msg)
             err = self._get_error_message_from_exception(e)
-            action_result.set_status(phantom.APP_ERROR, msg, err)
-            return action_result.get_status()
+            return action_result.set_status(phantom.APP_ERROR, msg, err)
         if not isinstance(mapping, dict):
-            action_result.set_status(phantom.APP_ERROR, 'Invalid JSON string. Must be a dictionary containing key-value pairs')
-            return action_result.get_status()
+            return action_result.set_status(phantom.APP_ERROR, 'Invalid JSON string. Must be a dictionary containing key-value pairs')
```
[the remark this answers is not in the document]
Comment (dhwanis-crest) [reply archer_connector.py:343]: Done

File: archer.json [hunk 5]
```diff
@@ -1105,18 +1434,417 @@
...
+                "type": "table",
+                "title": "Archer attachment",
+                "width": 1
+            },
+            "versions": "EQ(*)"
+        },
+        {
+            "action": "get report",
+            "description": "Get a list of tickets from a report",
+            "verbose": "The records for a report GUID (guid) are returned.Per page, Archer returns 50 records. Here the behavior of max_pages and max_results would be such that, if max_pages = 1 and max_results = 100, then the action would fetch only 50 records i.e. 1st page. If max_pages = 1 and max_results = 10, the action will return 10 records based on the max_results parameter. Also, the number of columns and record search depends on the columns displayed in reports on the Archer instance's UI, i.e if on UI the \"Summary\" column is not added to visible columns, it won't be displayed in action output as well as no records will be fetched if used in the filter parameters.",
```
[the remark this answers is not in the document]
Comment (dhwanis-crest) [reply archer.json:1533]: Done

File: archer_utils.py (hunk 1 above)
Comment (dhwanis-crest) [reply archer_utils.py:760]: Done

File: archer_utils.py [hunk 6]
```diff
@@ -683,32 +701,255 @@ def update_record(self, app, contentId, fieldId, value, doit=True):
...
-            result = ''
-        else:
-            result = ' - Terminate session details: {}'.format(result)
+    def merge_field_defs(self, field_defs, raw_records, max_count, total_count, page_number):
 
-        return result
+        try:
+
+            # Initialize result dictionary
+            merge_dict = {}
```
[the remark this answers is not in the document]
Comment (dhwanis-crest) [reply archer_utils.py:860]: done

File: archer_soap.py [hunk 7]
```diff
@@ -375,20 +392,55 @@ def _generate_xml_stub(self):
...
+            '/soap:Envelope/soap:Body/dummy:SearchRecordsByReportResponse/dummy:SearchRecordsByReportResult', namespaces=ALL_NS_MAP)
+
+        if len(rec_xml) > 0:
+            return {'status': 'success', 'result': rec_xml[0].text}
+        else:
+            rec_xml = resp_root.xpath('//*[local-name()="faultstring"]', namespaces=ALL_NS_MAP)
+            if len(rec_xml) > 0:
+                return {'status': 'failed', 'result': rec_xml[0].text}
+            else:
+                return {'status': 'failed', 'result': 'Unable to find SearchRecordsByReportResult.'}
```
[the remark this answers is not in the document]
Comment (dhwanis-crest) [reply archer_soap.py:418]: done

File: archer_consts.py [hunk 8]
```diff
@@ -12,34 +12,47 @@
...
 # and limitations under the License.
-ARCHER_ACTION_CREATE_TICKET = "create_ticket"
-ARCHER_ACTION_UPDATE_TICKET = "update_ticket"
-ARCHER_ACTION_GET_TICKET = "get_ticket"
-ARCHER_ACTION_LIST_TICKET = "list_tickets"
-ARCHER_ACTION_ON_POLL = "on_poll"
+ARCHER_ACTION_UPDATE_TICKET = 'update_ticket'
+ARCHER_ACTION_CREATE_TICKET = 'create_ticket'
+ARCHER_ACTION_GET_TICKET = 'get_ticket'
+ARCHER_ACTION_LIST_TICKET = 'list_tickets'
```
[the remark this answers is not in the document]
Comment (dhwanis-crest) [reply archer_consts.py:18]: done

File: archer_utils.py [hunk 9]
```diff
@@ -92,67 +92,71 @@ def __init__(self, base_url, userName, password, instanceName, usersDomain):
...
-                error_code = "Error code unavailable"
-                error_msg = "Error message unavailable. Please check the asset configuration and|or action parameters."
-        except:
-            error_code = "Error code unavailable"
-            error_msg = "Error message unavailable. Please check the asset configuration and|or action parameters."
-
-        if error_code in "Error code unavailable":
-            error_text = "Error Message: {0}".format(error_msg)
+        except Exception as e:
+            self.error_print('Error occurred while fetching exception information. Details: {}'.format(str(e)))
```
[the remark this answers is not in the document]
Comment (dhwanis-crest) [reply archer_utils.py:126]: As this is legacy code we have kept this for consistency.

File: archer_soap.py [hunk 10]
```diff
@@ -92,52 +93,45 @@ def _domain_user_authenticate(self):
...
-                    for node in name_ele.itersiblings(tag='Id'):
-                        return int(node.text)
+        try:
+            result = resp_root.xpath(archer_consts.ARCHER_XPATH_GROUP, namespaces=ALL_NS_MAP)
+            if result:
+                for name_ele in result:
+                    if name_ele.text == groupname:
+                        for node in name_ele.itersiblings(tag='Id'):
+                            return int(node.text)
+        except Exception:
```
[the remark this answers is not in the document]
Comment (dhwanis-crest) [reply archer_soap.py]: etree.XPathEvalError and AttributeError. Have made this change.

File: archer_utils.py [hunk 11]
```diff
@@ -162,11 +166,11 @@ def get_fieldId_for_content_and_name(self, cid, fname):
         """
         W('Getting fieldId for {} in record {}'.format(fname, cid))
         j = json.loads(self._rest_call('/api/core/content/{}'.format(cid)))
-        if not('RequestedObject' in j and 'FieldContents' in j['RequestedObject']):
+        if not ('RequestedObject' in j and 'FieldContents' in j['RequestedObject']):
```
[the remark this answers is not in the document]
Comment (dhwanis-crest) [reply archer_utils.py:169]: Done

File: archer_utils.py [hunk 12]
```diff
@@ -332,22 +335,29 @@ def get_valuesetvalue_of_field(self, fieldId, value):
...
+            if value and isinstance(value, str):
+                value = [x.strip() for x in value.split(",")]
+            user_id = []
+            group_id = []
+            for val in value:
+                uid = self.asoap.find_user(val)
+                gid = self.asoap.find_group(val)
+                group_id.append(gid)
+                user_id.append(uid)
+                if not (uid or gid):
```
[the remark this answers is not in the document]
Comment (dhwanis-crest) [reply archer_utils.py]: Done

File: archer_utils.py [hunk 13]
```diff
@@ -549,7 +570,7 @@ def find_records(self, app, field_name, value, max_count, comparison=None, sort=
                         if value_list:
                             if isinstance(value_list, dict):
                                 value_list = [value_list]
-                            value_list = set([ x.get('#text', '') for x in value_list])
+                            value_list = set([x.get('#text', '') for x in value_list])
```
[the remark this answers is not in the document]
Comment (dhwanis-crest) [reply archer_utils.py]: Done

File: archer_utils.py [hunk 14]
```diff
@@ -661,11 +675,15 @@ def update_record(self, app, contentId, fieldId, value, doit=True):
             doit, a boolean: whether to actually issue the update command.  If
                 False, return the data that would have been sent.
 
-            TODO: accept multiple field/value pairs for efficiency
         """
         W('In update_record({}, {}, {})'.format(contentId, fieldId, value))
         try:
```
[the remark this answers is not in the document]
Comment (dhwanis-crest) [reply archer_utils.py:688]: Have made these changes.

File: archer_views.py (hunk 2 above)
Comment (dhwanis-crest) [reply archer_views.py]: Done

File: archer_soap.py [hunk 15]
```diff
@@ -375,20 +392,55 @@ def _generate_xml_stub(self):
...
                 'Content-Type': 'text/xml; charset=utf-8',
-                'SOAPAction': '"http://archer-tech.com/webservices/{}"'.format(api),
+                'SOAPAction': '"http://archer-tech.com/webservices/{}"'.format(api_tag),
             }
             response = requests.post(  # nosemgrep: python.requests.best-practice.use-timeout.use-timeout
                 uri, data=xml, headers=headers, verify=self.verify_cert)
+            for x in archer_consts.ARCHER_INVALID_SESSION_TOKEN_MSG:
+                if x in response.text:
+                    generate_new_token = True
+                    break
```
[the remark this answers is not in the document]
Comment (dhwanis-crest) [reply archer_soap.py]: done

File: archer_utils.py [hunk 16]
```diff
@@ -638,14 +656,10 @@ def create_record(self, app, data={}):
...
             fields.append(field)
-
-        if not self.asoap:
-            self.asoap = ArcherSOAP(self.base_url, self.userName, self.password, self.instanceName,
-                     verify_cert=self.verifySSL, usersDomain=self.users_domain)
         cid = self.asoap.create_record(moduleId, fields)
         return cid
 
-    def update_record(self, app, contentId, fieldId, value, doit=True):
+    def update_record(self, app, contentId, fieldId, value, data={}, doit=True):
```
[the remark this answers is not in the document]
Comment (dhwanis-crest) [reply archer_utils.py:662]: yes, have made the changes

File: README.md [hunk 17]
```diff
@@ -115,7 +115,7 @@ Create a new ticket
...
+<p>JSON specifying the field names and values for a new Archer record (key/value pairs). For Cross-Reference fields, the value must be the content id of the referenced content.</p><p>Create record sample JSON: </p><pre><code>{ "Incident Summary": "test incident summary data", "Incident Owner": "testuser" }</code></pre><br><p>Parameter application is case-sensitive. The following field types are supported for creating a ticket:<ul><li>Type 1 (TextString)</li><li>Type 2 (Numeric)</li><li>Type 3 (Date with Time)</li><li>Type 4 (Values List)</li><li>Type 8 (Users/Groups List)</li><li>Type 9 (Cross-Reference)</li><li>Type 23 (Related Records)</li></ul></p>.
```
Comment (alexa-phantom) [inline README.md:118]: Why is a period added?

File: README.md [hunk 18]
```diff
@@ -141,7 +141,7 @@ Update the value of a field of a record
...
+There are multiple ways of locating a ticket to update. You must either give the content ID for the record, which can be obtained from Archer, or by specifying both the name of the Tracking ID field (name_field) and the Tracking ID (name_value). If all three parameters are provided, the content ID will be used as an overriding parameter to update the ticket. Parameters application, name_field, name_value, field_id, and value are case-sensitive. Here, if both json_string, and field_id and value are specified, preference would be given to json_string parameter and ticket will be updated based on that.  The following field types are supported for creating a ticket:<ul><li>Type 1 (TextString)</li><li>Type 2 (Numeric)</li><li>Type 3 (Date with Time)</li><li>Type 4 (Values List)</li><li>Type 8 (Users/Groups List)</li><li>Type 9 (Cross-Reference)</li><li>Type 23 (Related Records)</li></ul>.
```
Comment (alexa-phantom) [inline README.md:144]: Why is a period added?

File: README.md (hunk 17 above)
Comment (dhwanis-crest) [reply README.md:118]: We have a static check for verbose ending with period which is why we have made these changes.

**Comment by bb-splunk** [comment]
Have any tests been added/are planned for this change?

File: archer_connector.py [hunk 19]
```diff
@@ -464,49 +470,469 @@ def _validate_integer(self, action_result, parameter, key, allow_zero=False):
...
+
+        sec_alert_request = {}
+        sec_alert_request['Content'] = contentDetails
+
+        # make REST call
+        try:
+            r = archer_utils.ArcherAPISession._rest_call(self.proxy, consts.ARCHER_UPDATE_CONTENT_ENDPOINT, 'put', sec_alert_request)
+            r = json.loads(r)
+        except Exception:
+            return action_result.set_status(phantom.APP_ERROR, consts.ARCHER_ERR_ACTION_EXECUTION.format(self.get_action_identifier(), r))
```
Comment (bb-splunk) [inline archer_connector.py:923]: Same as above

**Review by bb-splunk — APPROVED** [review]

**Review by alexa-phantom — APPROVED** [review]
Please address comments from @bb-splunk

File: .pre-commit-config.yaml [hunk 20]
```diff
@@ -1,6 +1,6 @@
 repos:
```
[the remark this answers is not in the document]
Comment (dhwanis-crest) [reply .pre-commit-config.yaml:1]: Yes, we have added tests for the new and updated actions in App Tests gitlab repo, which are triggered whenever a new commit is added to the PR, also the playbooks have been added for the same in the Apps test Playbooks gitlab repo.

File: archer_connector.py [hunk 21]
```diff
@@ -464,49 +470,469 @@ def _validate_integer(self, action_result, parameter, key, allow_zero=False):
...
+        if results_filter_operator == 'and':
+            and_dict_len = len(results_filter_dict)
+            for record in records:
+                and_dict_count = 0
+                for field in record['Field']:
+                    for k, v in results_filter_dict.items():
+                        try:
+                            if results_filter_equality == 'equals':
+                                if field['@name'] == k and v.lower() == field['#text'].lower():
+                                    and_dict_count = and_dict_count + 1
```
[the remark this answers is not in the document]
Comment (dhwanis-crest) [reply archer_connector.py:616]: Done

File: archer_connector.py [hunk 22]
```diff
@@ -464,49 +470,469 @@ def _validate_integer(self, action_result, parameter, key, allow_zero=False):
...
+                    results_filter_operator = "and"
+        if results_filter_equality:
+            if results_filter_equality.lower() not in consts.ARCHER_EQUALITY_VALUELIST:
+                return action_result.set_status(phantom.APP_ERROR,
+                    f'Please enter a valid value for results_filter_equality from {consts.ARCHER_EQUALITY_VALUELIST}')
+            else:
+                results_filter_equality = results_filter_equality.lower()
+
+        status, max_count = self._validate_integer(action_result, max_count, 'max_result', False)
+        if (phantom.is_fail(status)):
```
[the remark this answers is not in the document]
Comment (dhwanis-crest) [reply archer_connector.py:694]: Done

File: archer_connector.py [hunk 23]
```diff
@@ -464,49 +470,469 @@ def _validate_integer(self, action_result, parameter, key, allow_zero=False):
...
+            msg = consts.ARCHER_ERR_VALID_JSON
+            self.debug_print(msg)
+            err = self._get_error_message_from_exception(e)
+            return action_result.set_status(phantom.APP_ERROR, msg, err)
+        if results_filter_dict and not isinstance(results_filter_dict, dict):
+            return action_result.set_status(phantom.APP_ERROR, consts.ARCHER_INVALID_JSON)
+
+        results_filter_operator = param.get('results_filter_operator')
+        results_filter_equality = param.get('results_filter_equality')
+        if results_filter_operator:
```
[the remark this answers is not in the document]
Comment (dhwanis-crest) [reply archer_connector.py:506]: Have made the changes.

File: archer_connector.py (hunk 19 above)
Comment (dhwanis-crest) [reply archer_connector.py]: done

File: archer_connector.py [hunk 24]
```diff
@@ -58,37 +62,61 @@ def __init__(self):
...
                 elif len(e.args) == 1:
-                    error_code = consts.ARCHER_ERROR_CODE_UNAVAILABLE
                     error_msg = e.args[0]
-            else:
-                error_code = consts.ARCHER_ERROR_CODE_UNAVAILABLE
-                error_msg = consts.ARCHER_ERROR_CHECK_ASSET_CONFIG
-        except:
-            error_code = consts.ARCHER_ERROR_CODE_UNAVAILABLE
-            error_msg = consts.ARCHER_ERROR_CHECK_ASSET_CONFIG
+        except Exception as e:
```
[the remark this answers is not in the document]
Comment (dhwanis-crest) [reply archer_connector.py:117]: We have this function added to our best practices, so for making this change we will have to test it thoroughly, so maybe we can incorporate these changes in future releases.

**Pull request merged**